File: frontend/src/lib/api/sso.ts
import apiClient from './client'

export type IdpType = 'saml' | 'oidc'

export interface SSOConfig {
  id: string
  tenant_id: string
  idp_type: IdpType
  entity_id: string
  sso_url: string
  attr_mapping: Record<string, unknown> | null
  is_active: boolean
  created_at: string
}

export interface SSOConfigCreate {
  idp_type: IdpType
  entity_id: string
  sso_url: string
  certificate?: string
  attr_mapping?: Record<string, unknown>
  domains?: string[]
}

export interface SSOConfigUpdate {
  entity_id?: string
  sso_url?: string
  certificate?: string
  attr_mapping?: Record<string, unknown>
  is_active?: boolean
  domains?: string[]
}

export interface ScimToken {
  id: string
  tenant_id: string
  name: string
  expires_at: string | null
  is_active: boolean
  created_at: string
}

export interface ScimTokenCreated extends ScimToken {
  raw_token: string
}

export interface DomainHint {
  domain: string
  sso_config_id: string
}

export const ssoApi = {
  getConfig: () =>
    apiClient.get<SSOConfig>('/api/v1/sso/config').then((r) => r.data),

  createConfig: (data: SSOConfigCreate) =>
    apiClient.post<SSOConfig>('/api/v1/sso/config', data).then((r) => r.data),

  updateConfig: (data: SSOConfigUpdate) =>
    apiClient.put<SSOConfig>('/api/v1/sso/config', data).then((r) => r.data),

  deleteConfig: () =>
    apiClient.delete('/api/v1/sso/config'),

  getDomainHint: (domain: string) =>
    apiClient.get<DomainHint>(`/api/v1/sso/domain-hint/${domain}`).then((r) => r.data),

  listScimTokens: () =>
    apiClient.get<ScimToken[]>('/api/v1/sso/scim-tokens').then((r) => r.data),

  createScimToken: (data: { name: string; expires_at?: string }) =>
    apiClient.post<ScimTokenCreated>('/api/v1/sso/scim-tokens', data).then((r) => r.data),

  revokeScimToken: (id: string) =>
    apiClient.delete(`/api/v1/sso/scim-tokens/${id}`),
}

File: backend/app/schemas/sso.py
import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, field_validator


class IdpType(str, Enum):
    saml = "saml"
    oidc = "oidc"


# ── SSO Config ────────────────────────────────────────────────────────────────

class SSOConfigCreate(BaseModel):
    idp_type: IdpType
    entity_id: str
    sso_url: str
    certificate: str | None = None   # Raw PEM cert — encrypted at write time
    attr_mapping: dict | None = None
    domains: list[str] = []          # Email domains that trigger SSO auto-redirect


class SSOConfigUpdate(BaseModel):
    entity_id: str | None = None
    sso_url: str | None = None
    certificate: str | None = None
    attr_mapping: dict | None = None
    is_active: bool | None = None
    domains: list[str] | None = None


class SSOConfigRead(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    idp_type: str
    entity_id: str
    sso_url: str
    attr_mapping: dict | None
    is_active: bool
    created_at: datetime
    # certificate and domains are never returned in reads for security

    model_config = ConfigDict(from_attributes=True)


# ── SCIM Token ────────────────────────────────────────────────────────────────

class ScimTokenCreate(BaseModel):
    name: str
    expires_at: datetime | None = None


class ScimTokenRead(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    expires_at: datetime | None
    is_active: bool
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ScimTokenCreated(ScimTokenRead):
    raw_token: str  # shown only once at creation


# ── Domain hint ───────────────────────────────────────────────────────────────

class DomainHintRead(BaseModel):
    domain: str
    sso_config_id: uuid.UUID

    model_config = ConfigDict(from_attributes=True)


# ── SP Metadata ───────────────────────────────────────────────────────────────

class SPMetadata(BaseModel):
    entity_id: str
    acs_url: str
    metadata_xml: str
